File: producer.py


import time
import pika


host = 'localhost'
port = 5672

exchange_name = "my-exchange"
queue_name = "my-queue"

# connect to RabbitMQ server and create a channel
connection_params = pika.ConnectionParameters(host=host, port=port)
connection = pika.BlockingConnection(connection_params)

channel = connection.channel()

# create an exhange
channel.exchange_declare(exchange=exchange_name, exchange_type='fanout')

# create a queue
queue = channel.queue_declare(queue=queue_name)

# bind the exchange and queue
channel.queue_bind(exchange=exchange_name, queue=queue_name)

print("Input messages to produce or `stop` to stop")

while True:
	message = input("Enter message to send or stop to exit: ")

	channel.basic_publish(exchange=exchange_name, routing_key='', body=message)
	print("message published")

	if message.lower() == "stop":
		break

# sleep for 2s waiting for consumer to receive stop message and delete queue
time.sleep(2)
channel.queue_delete(queue_name)
channel.exchange_delete(exchange_name)


connection.close()

File: producer.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
)

var host string = "localhost"
var port int = 5672

var queue_name string = "my-queue"
var exchange_name string = "my-exchange"

func main() {

	// connect to rabbitMQ
	conn, _ := amqp.Dial("amqp://guest:guest@localhost:5672/")

	// open a channel
	ch, _ := conn.Channel()

	// declare queue
	ch.QueueDeclare(
		queue_name, // name
		false,      // durable
		false,      // delete when unused
		false,      // exclusive
		false,      // no-wait
		nil,        // arguments
	)

	// declare an exchange
	ch.ExchangeDeclare(
		exchange_name, // name
		"fanout",      // exchange type
		false,         // durable
		false,         // delete when unused
		false,         // exclusive
		false,         // no-wait
		nil,           // arguments
	)

	ch.QueueBind(
		queue_name,    // destination
		"",            // key
		exchange_name, // source
		false,         // nowait
		nil,           // args
	)

	reader := bufio.NewReader(os.Stdin)

	for true {

		fmt.Print("Enter message to publish or stop exit: ")
		message, _ := reader.ReadString('\n')

		m := amqp.Publishing{
			ContentType: "text/plain",
			Body:        []byte(message),
		}

		ch.Publish(
			exchange_name, // exchange name
			queue_name,    // key
			false,         // mandatory
			false,         // immediate
			m,             // message body
		)

		if strings.HasPrefix(message, "stop") {
			break
		}

		time.Sleep(time.Millisecond * 10)
	}

	// delete queue
	ch.QueueDelete(
		queue_name, // queue name
		false,      // ifunused
		false,      // ifempty
		false,      // nowait
	)

	// delete channel
	ch.ExchangeDelete(
		exchange_name, // exchange name
		false,         // ifunused
		false,         // nowait
	)

	// close connection
	defer conn.Close()
}
